### PR TITLE
CMT straxen implementation

### PR DESCRIPTION
**Pull request by ahiguera-mx**
**What is the problem / what does the code in this PR do**
CMT implementation in straxen
**Can you briefly describe how it works?**

Corrections are a set of parameters to be applied at the event builder process to remove detectors effects, corrections are stored as pandas.DataFrame with a DatetimeIndex and housed at mongoDB. A general implementation about CMT can be found at strax, see for instance https://github.com/AxFoundation/strax/blob/master/strax/corrections.py and https://github.com/XENONnT/analysiscode/blob/master/StraxTests/corrections_strax_test.ipynb

**Can you give a minimal working example (or illustrate with a figure)?**
Please see this notebook, https://github.com/ahiguera-mx/xenon/blob/master/corrections_straxen_test.ipynb
If necessary I can move this to analysiscode/StraxTest
TODO: update contexts according to AC input
Please include the following if applicable:
  - Update the docstring(s)
  - Update the documentation
  - Tests to check the (new) code is working as desired.
  - Does it solve one of the open issues on github?

Please make sure that all automated tests have passed before asking for a review (you can save the PR as a draft otherwise).

### CODE REVIEW

**Comment by ahiguera-mx** [comment]
FYI @jorana @WenzDaniel the travis test failed because could not find xenon1t_secrest.py which I expected to happened since this is private info

**Comment by tunnell** [comment]
@ahiguera-mx for the runDB, can you hard code it in the test?  Specifically, TravisCI lets you launch a fake MongoDB.  I don't think we want the tests to have the real MongoDB password.  I think you add "services: -mongodb" to .travis.yml.  Then it's just local access and decouples straxen from the RunDB.

File: straxen/corrections_services.py
Comment (WenzDaniel) [inline straxen/corrections_services.py:71]: In the doc string it says dict?

File: straxen/corrections_services.py
Comment (WenzDaniel) [inline straxen/corrections_services.py:73]: This looks like it must be a tuple, but I think you are right in the doc-string a dict would be a more elegant solution here. 

You could test if the keys are specified and if not add the corresponding thing as a default e.g.

```
if not xeno1t in config_model.keys():
    config_model[xenon1t] = Flase
```

File: straxen/corrections_services.py
Comment (WenzDaniel) [inline straxen/corrections_services.py:106]: Hej Aaron, I looked through `get_elife` and `get_pmt_gains` and they both look very very similar. I think this will also be the case for the not yet implemented functions like `get_lce` and `get_fdc` so I was wondering if we could not be a bit smarter and just have a single function for all of those to avoid this copy and paste code. 
At the moment I do not see why this should not be possible. In case we have to change something on the straxen side (e.g. the format strax expects the PMT gains to be) no problem we can do that.

File: straxen/corrections_services.py
Comment (WenzDaniel) [inline straxen/corrections_services.py:183]: Maybe we should make this function stand alone and put it into common.py since also other projects could benefit from it, what do you think?

**Review by WenzDaniel — CHANGES_REQUESTED** [review]
Hej Aaron, I had a first superficial reading of your code. I think in general it is fine, though I think we can do better with the functions called `get_*`. Could we maybe discuss and see if we can unify those functions such that we do not have to copy and paste a 95% identical code for each correction?

**Review by jorana — CHANGES_REQUESTED** [review]
Hi Aaron,
Thanks for the PR. Let me start with the broader issues and comment on small code/style stuff with the comments below.

Things to consider first
- We might want to consider smart cashing of the results if we want to allow users to query the database in a for-loop (as will be possible with the tool at the moment). @darrylmasson will have comments on this. 
- We need to be smart about the write/read permissions. Few should have write access and the tool should be keeping this in mind. We do this for instance in the RunDB as well.
- Can you move the init of the runsb to the __init__ of your class to make the initialization of the rundb more clear and only once. Otherwise your initializing it every time again you are using one of the functions.
- If I run your notebook, some of the queries take a lot of time. Do you have the same? If so you might want to use a profiler to check which functions use all of the time. There might be some over-redundancy that takes a long time. For example:
```python
%%time
elife_model=('elife_model','v1',True)
cmt.get_corrections_config('170204_1710','elife',elife_model)
#gives:
CPU times: user 969 ms, sys: 31.2 ms, total: 1 s
Wall time: 3.34 s
...
```

In just a moment I'll also add some comments to the code itself.

**Review by jorana — CHANGES_REQUESTED** [review]

File: straxen/corrections_services.py
Comment (darrylmasson) [reply straxen/corrections_services.py:217]: Or, `rundoc = collection.find_one({'name' if xenon1t else 'number': run_id, 'end' : {'$exists': 1}}, {'start': 1})`, if no doc is found then it returns `None`

**Comment by darrylmasson** [comment]
Maybe I'm too late to this party, and maybe this PR isn't the best place to raise this particular concern, but I have some misgivings about the database structure behind this. The gains of each PMT are stored in a separate collection, which means we have ~750 collections of gains between 1T and nT. To load the gains (which can happen many times because strax is quite liberal with instantiating plugins), the CMT first queries the database to see which model(s) to use, then queries again individually for each PMT. This is _500 separate queries per plugin instance per run_, which feels monumentally inefficient to me. Even if you cache, that just normalizes out the "per plugin" bit, and you're still left with hundreds of queries which could in principle (_waves hands_) come from a single aggregate call if things were in one collection.

**Comment by tunnell** [comment]
> Maybe I'm too late to this party, and maybe this PR isn't the best place to raise this particular concern, but I have some misgivings about the database structure behind this. The gains of each PMT are stored in a separate collection, which means we have ~750 collections of gains between 1T and nT. To load the gains (which can happen many times because strax is quite liberal with instantiating plugins), the CMT first queries the database to see which model(s) to use, then queries again individually for each PMT. This is _500 separate queries per plugin instance per run_, which feels monumentally inefficient to me. Even if you cache, that just normalizes out the "per plugin" bit, and you're still left with hundreds of queries which could in principle (_waves hands_) come from a single aggregate call if things were in one collection.

This was my idea so I'll chime in.  Specifically, we can always do it with one aggregation query that also selects only a time range of interest later on.  I thought we should prioritize getting the key functionality in, then optimizing (premature optimization root of all evil).  Such a simple structure will make it easier for others to interface, while also opening up aggreations that allow running over multiple collections.

**Comment by darrylmasson** [comment]
Premature optimization is a very slippery slope I have fallen down before, this is true. However, I disagree that this structure makes it easy for others to interface, because only a very few will be accessing this directly via the database. The vast majority will access via the CMT API and neither know nor care about the underlying structure. Also, I shudder to think of the aggregate command that first starts with 500 `$lookup`s, and I feel like indexing 500 different collections is rather less efficient than indexing 1.

**Comment by tunnell** [comment]
The PMT group wants to treat them separately because also it becomes a nightmare to manage (remember the giant sympy before)?  I'll try later today to write some aggregation for it.  Most collections will be cached and have just 2 or 3 entries.  Aggreagrations allow you to output the result to another collection, that can have the structure you mention, but you want mongo to generate it automatically and the PMT group to be able to look at easy tables.

**Comment by tunnell** [comment]
Also, @darrylmasson strax chat?  Since we do have corrections anyways and this is a cleanup, there's no internal XENON info being released here related to our schedule. However, it's easier through Slack.

File: straxen/plugins/event_processing.py
Comment (ahiguera-mx) [reply straxen/plugins/event_processing.py:368]: yes, thank you

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:71]: it as tuple

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:73]: could be a more elegant solution, I will revise this later as Chris says "premature optimization root of all evil"

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:34]: thanks, fixed

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:104]: fixed

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:216]: yes, I meant that, it is just a place holder for having a date time format

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:173]: fixed

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:135]: fixed

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:188]: fixed

File: straxen/plugins/event_processing.py
Comment (ahiguera-mx) [reply straxen/plugins/event_processing.py:368]: fixed

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:201]: @jorana can you elaborate more here, not sure I understand the comment, should I remove the requirement of detectors?

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:220]: fixed

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:201]: never mind, fixed see below

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:217]: thanks, indeed that works

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:208]: fixed

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:194]: I agree that this part should be in the init, I have done that
Regarding using the RunDB class, then I would have the same issue, every time I need to get the start time of a run, I will call this class, then this class would initialize the runDB every time is called, right? or I missing something there 
I saw this class is used at the level of context... maybe there is some tricks there

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:114]: fixed

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:156]: fixed

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:19]: I did try to do that, could not make it work... probably this optimization we can revisit later

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:122]: see comment above

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:183]: can be a good idea, not sure how many would use it

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:19]: I think the reason being the child class(CmtServices) inherits all the methods and properties from its parent, but I'm initializing strax.CorrectionsInterface() by checking that any of the host is working, so super() try to inherits all, latter I'm using the methods from init from the parent class and this creates a conflict

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:89]: see above comment

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:88]: I tried to do something similar to what Jelle did here 
https://github.com/ahiguera-mx/straxen/blob/master/straxen/gain_models.py#L45
but probably I went too far. 
I have add a check for that

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:86]: what the number of pmts? not sure I understand the comment

File: straxen/plugins/event_processing.py
Comment (ahiguera-mx) [reply straxen/plugins/event_processing.py:338]: done

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py]: I have done that, thanks for the suggestion

**Comment by ahiguera-mx** [comment]
> Hi Aaron,
> Thanks for the PR. Let me start with the broader issues and comment on small code/style stuff with the comments below.
> 
> Things to consider first
> 
> * We might want to consider smart cashing of the results if we want to allow users to query the database in a for-loop (as will be possible with the tool at the moment). @darrylmasson will have comments on this.

This has been discussed between @tunnell and @darrylmasson, probably they should summarize the discussion

> * We need to be smart about the write/read permissions. Few should have write access and the tool should be keeping this in mind. We do this for instance in the RunDB as well.

Users have only read permission to any of the database used here, only correction manager has read/write permission 

> * Can you move the init of the runsb to the **init** of your class to make the initialization of the rundb more clear and only once. Otherwise your initializing it every time again you are using one of the functions.

Thanks, this was a good catch, I've done that, this speeds up things 

> * If I run your notebook, some of the queries take a lot of time. Do you have the same? If so you might want to use a profiler to check which functions use all of the time. There might be some over-redundancy that takes a long time. For example:

This is related to the point rise by Darryl, so he and Chris can elaborate more on that 
> 
> ```python
> %%time
> elife_model=('elife_model','v1',True)
> cmt.get_corrections_config('170204_1710','elife',elife_model)
> #gives:
> CPU times: user 969 ms, sys: 31.2 ms, total: 1 s
> Wall time: 3.34 s
> ...
> ```
> 
> In just a moment I'll also add some comments to the code itself.

**Comment by tunnell** [comment]
@ahiguera-mx the summary of the discussion with Darryl was that we can use MongoDB aggregations to do most of what your strax/straxen code is doing server side in under a second.  See here:

https://gist.github.com/tunnell/557afc9d05f0018f9609bb82dbeecda9
https://docs.mongodb.com/manual/aggregation/

However, it comes at the expense of usability since it's MongoDB aggregation voodoo.  Therefore, if we encounter speed issues (unlikely anytime soon) then we have a reasonable solution.   Therefore, I suggest we use your implementation for now, then (if required) we can always switch.

**Comment by tunnell** [comment]
@jorana I think the reason that some of those calculations are slow is because we have one 'correction point' per run for XENON1T in strax.  However, this is a full experimental exposure that is highly oversampled.  This is why I think we should consider merging this for now, then in a few years if we have slowness, we can run the code I linked above instead (which performs the calculation then puts in caching collection)

File: straxen/corrections_services.py
Comment (jorana) [reply straxen/corrections_services.py:17]: I don't see how this would be a problem. The point is exactly that you want to use those methods from the parent.

Comment (ahiguera-mx) [reply straxen/corrections_services.py:17]: The problems that I had were authentication issues that are probably related to mongo, not necessary with inherent a class

**Review by jorana — COMMENTED** [review]
Hi Aaron, thanks for the many changes. I will want to try some things and just open a PR to do so, do need some time for that.

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:39]: fixed

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:84]: fixed

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:193]: fixed

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:176]: fixed

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:101]: yes, it gives the same results and don't see any issues with the RAM

**Review by jorana — APPROVED** [review]
As far as I'm concerned this PR is ready for a testing phase / implementation (as soon as we have finished the XXX of this week). The only thing I still would like to add the `get_mongo_url` from https://github.com/XENONnT/straxen/pull/216 but that’s better after #216 is merged.

The testing should commence before merging, as soon as we merge the CMT is automatically replacing the previous implementations, as such please wait with the merging.

**Comment by jorana** [comment]
In https://github.com/ahiguera-mx/straxen/pull/3 I've done what I mentioned above. I don't have any pending comments at the moment.

@WenzDaniel will be looking at this PR in the comming days if I'm correct.

**Review by WenzDaniel — APPROVED** [review]
Hi Aaron, thanks a lot for your awesome work. You improved a lot since your first PR and for a Python beginner it is already very nice code. I just have some minor comments which should be fast to be implemented.

File: straxen/corrections_services.py
Comment (jorana) [reply straxen/corrections_services.py:185]: Good idea, so:
```n_tpc_pmts  = x1t_common_config['n_tpc_pmts ']```
I made the todo but indeed, this is a nice solution!

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:142]: done

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:152]: done

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:125]: it is part of `corrections.py` in strax

File: straxen/corrections_services.py
Comment (ahiguera-mx) [reply straxen/corrections_services.py:185]: Great, I did the change, now is just hard coded in `x1t_common_config` :)

File: straxen/corrections_services.py
Comment (jorana) [reply straxen/corrections_services.py:126]: I think that Daniel comments on the 'how'  argument that is now referring to the `interpolate`  method without using providing the `ffil` method as an option. I.e. that you always use this line:
https://github.com/ahiguera-mx/strax/blob/90b853562227b8ca28bb529a728baeb27d491e63/strax/corrections.py#L97

I think that it was agreed to use this `interpolation` method so I don't think it's much of a problem that it's not explicitly called again as a kwarg in this function (` df = self.interface.interpolate(df, when, how='interpolate')`). Also wouldn't hurt if it were.

**Pull request merged**
